pkgs/development/misc/resholved/mk-resholved-derivation.nix

```diff
@@ -0,0 +1,23 @@
+{ stdenv, lib, resholved, }:
+
+{ pname, src, version, scripts, inputs ? [ ], allow ? { }, passthru ? { }, ...
+}@attrs:
+let
+  inherit stdenv;
+  self = (stdenv.mkDerivation ((removeAttrs attrs [ "script" "inputs" "allow" ])
+    // {
+      inherit pname version src;
+      buildInputs = [ resholved ];
+      # tentatively disabled because gc probably knows things I don't :)
+      # propagatedBuildInputs = inputs;
```

Comment:
I've tentatively disabled this after graham commented on IRC:

> gchristensen | hmm yeah I'm not sure why this is a thing:       propagatedBuildInputs = inputs;

I think he's right--this is vestigial--but I wanted to comment on it here until this is reviewed since this is something I wasn't sure how to handle in the first place.

---
Since the first working draft, resholved has followed source statements and resolved them along with the parent script. This behavior is important--it helps resholved weed functions from sourced libraries out of the list of commands it needs to resolve. The naive first version needed to propagate inputs to ensure they were available while resolving a consumer.

A later update added the tail "directive" comments that encode this resolve-time state for later reference, making it no longer necessary to propagate the inputs in order to resolve shell libraries.

---

While reasoning through this, I realized there's a related question about how to square shell norms and conventions with Nix norms and conventions. Some executables used in shell scripts/libraries are coincidental implementation details that a consumer probably shouldn't rely on (and their dependencies not propagated)--but sometimes they are integral and I'm a little less certain about whether they should come along?

Take, for example, two fairly popular bash dotfile managers: homeschick and yadm. Both of these programs are fairly thin wrappers around git, with varying degrees of reliance on the user also running bare git commands. I don't know what razor to use to decide if git should be propagated along with things like this?

Reply:
I would err on the side of not propagating whenever it isn't required to make the software work. Explicit environment specification is a great thing, and propagation is a hack that should be avoided whenever possible; in an ideal world we wouldn't need it at all. (I can imagine cases where you'd want to deploy dotfiles from git without wanting to do actual commits and repository manipulation on that machine, for instance, though I admit this specific example isn't hugely convincing.)